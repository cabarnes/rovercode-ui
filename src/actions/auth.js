// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const UPDATE_VALID_AUTH = 'UPDATE_VALID_AUTH';
export const USER_LOGOUT = 'USER_LOGOUT';

// action creators
export const updateValidAuth = (isValidAuth) => ({
  type: UPDATE_VALID_AUTH,
  payload: isValidAuth,
});

export const logout = () => ({
  type: USER_LOGOUT,
});

// helper functions
export const checkAuthError = (dispatch) => (error) => {
  if (error.response && error.response.status === 401) {
    // Authentication is no longer valid
    dispatch(updateValidAuth(false));
  } else {
    throw error;
  }
};

export const authHeader = (cookies) => ({
  headers: {
    Authorization: `JWT ${cookies.get('auth_jwt')}`,
  },
});
