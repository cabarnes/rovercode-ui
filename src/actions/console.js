// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const APPEND = 'APPEND';
export const CLEAR = 'CLEAR';

// action creators
export const append = (message) => ({
  type: APPEND,
  payload: message,
});

export const clear = () => ({
  type: CLEAR,
});
