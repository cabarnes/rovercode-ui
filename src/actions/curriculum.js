// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

import axios from 'axios';

export const FETCH_COURSES = 'FETCH_COURSES';
export const FETCH_COURSES_PENDING = `${FETCH_COURSES}_PENDING`;
export const FETCH_COURSES_FULFILLED = `${FETCH_COURSES}_FULFILLED`;
export const FETCH_COURSES_REJECTED = `${FETCH_COURSES}_REJECTED`;

// action creators
export const fetchCourses = (xhrOptions) => ({
  type: FETCH_COURSES,
  meta: {
    debounce: {
      time: 500,
    },
  },
  payload: axios.get('/api/v1/courses/', xhrOptions)
    .then(({ data }) => (
      data
    )),
});
