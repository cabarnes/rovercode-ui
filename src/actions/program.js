// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

import axios from 'axios';

export const CLEAR_PROGRAMS = 'CLEAR_PROGRAMS';
export const FETCH_PROGRAMS = 'FETCH_PROGRAMS';
export const FETCH_PROGRAMS_PENDING = `${FETCH_PROGRAMS}_PENDING`;
export const FETCH_PROGRAMS_FULFILLED = `${FETCH_PROGRAMS}_FULFILLED`;
export const FETCH_PROGRAMS_REJECTED = `${FETCH_PROGRAMS}_REJECTED`;
export const REMOVE_PROGRAM = 'REMOVE_PROGRAM';
export const REMOVE_PROGRAM_PENDING = `${REMOVE_PROGRAM}_PENDING`;
export const REMOVE_PROGRAM_FULFILLED = `${REMOVE_PROGRAM}_FULFILLED`;
export const REMOVE_PROGRAM_REJECTED = `${REMOVE_PROGRAM}_REJECTED`;

// action creators
export const clearPrograms = () => ({
  type: CLEAR_PROGRAMS,
});

export const fetchPrograms = (xhrOptions) => ({
  type: FETCH_PROGRAMS,
  meta: {
    debounce: {
      time: 500,
    },
  },
  payload: axios.get('/api/v1/block-diagrams/', xhrOptions)
    .then(({ data }) => (
      data
    )),
});

export const removeProgram = (id, xhrOptions) => ({
  type: REMOVE_PROGRAM,
  payload: axios.delete(`/api/v1/block-diagrams/${id}/`, xhrOptions)
    .then(({ data }) => (
      data
    )),
});
