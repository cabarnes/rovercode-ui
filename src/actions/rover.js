// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const SCAN = 'SCAN';
export const SCAN_FULFILLED = `${SCAN}_FULFILLED`;
export const SCAN_REJECTED = `${SCAN}_REJECTED`;
export const CONNECT_ROVER = 'CONNECT_ROVER';
export const CONNECT_ROVER_FULFILLED = `${CONNECT_ROVER}_FULFILLED`;
export const CONNECT_ROVER_REJECTED = `${CONNECT_ROVER}_REJECTED`;
export const SEND_ROVER = 'SEND_ROVER';
export const SEND_ROVER_FULFILLED = `${SEND_ROVER}_FULFILLED`;
export const SEND_ROVER_REJECTED = `${SEND_ROVER}_REJECTED`;
export const DISCONNECT_ROVER = 'DISCONNECT_ROVER';

const UART_SERVICE_UUID = '6e400001-b5a3-f393-e0a9-e50e24dcca9e';
const UART_TX_CHARACTERISTIC_UUID = '6e400002-b5a3-f393-e0a9-e50e24dcca9e';
const UART_RX_CHARACTERISTIC_UUID = '6e400003-b5a3-f393-e0a9-e50e24dcca9e';

// action creators
export const scan = () => ({
  type: SCAN,
  payload: navigator.bluetooth.requestDevice({
    filters: [{ namePrefix: 'BBC micro:bit' }],
    optionalServices: [UART_SERVICE_UUID],
  }),
});

export const connect = (device, onMessage) => ({
  type: CONNECT_ROVER,
  payload: device.gatt.connect()
    .then(server => server.getPrimaryService(UART_SERVICE_UUID))
    .then(service => Promise.all([
      service.getCharacteristic(UART_RX_CHARACTERISTIC_UUID),
      service.getCharacteristic(UART_TX_CHARACTERISTIC_UUID)
        .then((characteristic) => {
          characteristic.startNotifications();
          characteristic.addEventListener(
            'characteristicvaluechanged',
            onMessage,
          );
          return characteristic;
        }),
    ])),
});

export const send = (channel, message) => ({
  type: SEND_ROVER,
  payload: channel.writeValue(message),
});

export const disconnect = device => ({
  type: DISCONNECT_ROVER,
  payload: device.gatt.disconnect(),
});
