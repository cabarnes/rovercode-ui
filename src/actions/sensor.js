// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

export const CHANGE_LEFT_SENSOR_STATE = 'CHANGE_LEFT_SENSOR_STATE';
export const CHANGE_RIGHT_SENSOR_STATE = 'CHANGE_RIGHT_SENSOR_STATE';
export const CHANGE_LIGHT_SENSOR_READINGS = 'CHANGE_LIGHT_SENSOR_READINGS';

// Sensor States
export const COVERED = 1;
export const NOT_COVERED = 2;

// action creators
export const changeLeftSensorState = (state) => ({
  type: CHANGE_LEFT_SENSOR_STATE,
  payload: state,
});

export const changeRightSensorState = (state) => ({
  type: CHANGE_RIGHT_SENSOR_STATE,
  payload: state,
});

export const changeLightSensorReadings = (leftReading, rightReading) => ({
  type: CHANGE_LIGHT_SENSOR_READINGS,
  payload: { leftReading, rightReading },
});
