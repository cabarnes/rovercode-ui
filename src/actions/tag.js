// actions https://redux.js.org/basics/actions
// async actions https://redux.js.org/advanced/asyncactions

import axios from 'axios';

export const FETCH_TAGS = 'FETCH_TAGS';
export const FETCH_TAGS_PENDING = `${FETCH_TAGS}_PENDING`;
export const FETCH_TAGS_FULFILLED = `${FETCH_TAGS}_FULFILLED`;
export const FETCH_TAGS_REJECTED = `${FETCH_TAGS}_REJECTED`;

// action creators
export const fetchTags = (xhrOptions) => ({
  type: FETCH_TAGS,
  payload: axios.get('/api/v1/tags/', xhrOptions)
    .then(({ data }) => (
      data
    )),
});
