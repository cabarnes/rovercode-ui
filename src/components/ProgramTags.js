import React, { Component } from 'react';
import { Chip, Grid, TextField } from '@material-ui/core';
import { Autocomplete } from '@material-ui/lab';
import { connect } from 'react-redux';
import { hot } from 'react-hot-loader';
import { withCookies } from 'react-cookie';
import { injectIntl } from 'react-intl';
import PropTypes from 'prop-types';

import { checkAuthError, authHeader } from '@/actions/auth';
import { changeProgramTags as actionChangeProgramTags } from '@/actions/code';
import { fetchTags as actionFetchTags } from '@/actions/tag';

const mapStateToProps = ({ code, tag }) => ({ code, tag });
const mapDispatchToProps = (dispatch, { cookies }) => ({
  fetchTags: () => dispatch(actionFetchTags(authHeader(cookies))).catch(checkAuthError(dispatch)),
  changeProgramTags: (id, tags) => dispatch(actionChangeProgramTags(id, tags, authHeader(cookies)))
    .catch(checkAuthError(dispatch)),
});

class ProgramTags extends Component {
  constructor(props) {
    super(props);

    this.state = {
      localTags: [],
    };
  }

  componentDidMount() {
    const { code, fetchTags } = this.props;

    fetchTags();

    this.setState({
      localTags: code.tags,
    });
  }

  handleChange = (event, value) => {
    this.setState({
      localTags: value,
    }, () => this.handleSave());
  }

  handleSave = () => {
    const { changeProgramTags, code } = this.props;
    const { localTags } = this.state;

    changeProgramTags(code.id, localTags);
  }

  render() {
    const { code, tag, intl } = this.props;
    const { localTags } = this.state;

    const tagPlaceholder = intl.formatMessage({
      id: 'app.program_tags.placeholder',
      description: 'Placeholder for tags input',
      defaultMessage: 'Add tags...',
    });

    return (
      <Grid container direction="column" justify="center" alignItems="stretch" spacing={1}>
        <Grid item>
          <Autocomplete
            id="tag-select"
            multiple
            freeSolo
            filterSelectedOptions
            size="small"
            options={tag.tags.map((globalTag) => globalTag.name)}
            onChange={this.handleChange}
            value={localTags}
            disabled={code.isReadOnly}
            renderInput={(params) => (
              <TextField {...params} variant="outlined" label={tagPlaceholder} />
            )}
            renderTags={(value, getTagProps) => value.map((option, index) => (
              <Chip color="secondary" size="small" label={option} {...getTagProps({ index })} />
            ))}
          />
        </Grid>
      </Grid>
    );
  }
}

ProgramTags.defaultProps = {
  tag: {
    tags: [],
  },
};

ProgramTags.propTypes = {
  code: PropTypes.shape({
    id: PropTypes.number,
    tags: PropTypes.arrayOf(PropTypes.string),
    isReadOnly: PropTypes.bool,
  }).isRequired,
  tag: PropTypes.shape({
    tags: PropTypes.arrayOf(PropTypes.shape({
      name: PropTypes.string,
    })),
  }),
  intl: PropTypes.shape({
    formatMessage: PropTypes.func.isRequired,
  }).isRequired,
  changeProgramTags: PropTypes.func.isRequired,
  fetchTags: PropTypes.func.isRequired,
};

export default hot(module)(
  withCookies(
    injectIntl(
      connect(mapStateToProps, mapDispatchToProps)(ProgramTags),
    ),
  ),
);
