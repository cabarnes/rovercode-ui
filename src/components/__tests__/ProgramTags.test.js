import React from 'react';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import { Cookies } from 'react-cookie';

import { updateValidAuth } from '@/actions/auth';
import ProgramTags from '../ProgramTags';

jest.mock('@/actions/code');
jest.mock('@/actions/tag');

import { changeProgramTags } from '@/actions/code'; // eslint-disable-line import/first, import/order

const cookiesValues = { auth_jwt: '1234' };
const cookies = new Cookies(cookiesValues);

describe('The ProgramTags component', () => {
  const mockStore = configureStore();
  const context = { cookies };
  let store;

  beforeEach(() => {
    store = mockStore({
      code: {
        tags: ['tag1', 'tag2'],
        isReadOnly: false,
      },
      tag: {
        tags: [{
          name: 'tag1',
        }, {
          name: 'tag2',
        }, {
          name: 'tag3',
        }],
      },
    });
    store.dispatch = jest.fn().mockResolvedValue();
  });

  test('renders on the page with no errors', () => {
    const wrapper = mountWithIntl(<ProgramTags store={store} />, { context });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('displays tags', () => {
    const wrapper = shallowWithIntl(
      <ProgramTags store={store} />, { context },
    ).dive().dive().dive()
      .dive()
      .dive()
      .dive()
      .dive();

    expect(wrapper.find('WithStyles(WithStyles(ForwardRef(Autocomplete)))').props().value).toStrictEqual(['tag1', 'tag2']);
    expect(wrapper.find('WithStyles(WithStyles(ForwardRef(Autocomplete)))').props().disabled).toBe(false);
  });

  test('disabled when read only', () => {
    const localStore = mockStore({
      code: {
        name: 'test name',
        isReadOnly: true,
        tags: [],
      },
    });
    localStore.dispatch = jest.fn().mockResolvedValue();
    const wrapper = shallowWithIntl(
      <ProgramTags store={localStore} />,
      { context },
    ).dive().dive().dive()
      .dive()
      .dive()
      .dive()
      .dive();

    expect(wrapper.find('WithStyles(WithStyles(ForwardRef(Autocomplete)))').props().disabled).toBe(true);
  });

  test('handles change', () => {
    const wrapper = shallowWithIntl(
      <ProgramTags store={store} />, { context },
    ).dive().dive().dive()
      .dive()
      .dive()
      .dive()
      .dive();

    wrapper.find('WithStyles(WithStyles(ForwardRef(Autocomplete)))').simulate('change', {}, ['tag3']);
    wrapper.update();

    expect(wrapper.find('WithStyles(WithStyles(ForwardRef(Autocomplete)))').props().value).toStrictEqual(['tag3']);
    expect(store.dispatch).toHaveBeenCalled();
    expect(store.dispatch).toHaveBeenCalledWith(changeProgramTags(1, ['tag3']));
  });

  test('handles authentication error', (done) => {
    const error = new Error();
    error.response = {
      status: 401,
    };
    store.dispatch = jest.fn();
    store.dispatch.mockRejectedValueOnce(error);
    store.dispatch.mockResolvedValue();

    const wrapper = shallowWithIntl(<ProgramTags store={store} />, { context }).dive().dive().dive()
      .dive()
      .dive();
    wrapper.dive().props().changeProgramTags(1, ['tag1', 'tag2']).then(() => {
      expect(store.dispatch.mock.calls.length).toBe(2);
      expect(store.dispatch).toHaveBeenCalledWith(
        changeProgramTags(1, ['tag1', 'tag2'], {
          headers: {
            Authorization: `JWT ${cookiesValues.auth_jwt}`,
          },
        }),
      );
      expect(store.dispatch).toHaveBeenCalledWith(updateValidAuth(false));
      done();
    });
  });

  test('handles other error', (done) => {
    const error = new Error();
    error.response = {
      status: 500,
    };
    store.dispatch = jest.fn().mockRejectedValue(error);

    const wrapper = shallowWithIntl(<ProgramTags store={store} />, { context }).dive().dive().dive()
      .dive()
      .dive();
    wrapper.dive().props().changeProgramTags(1, ['tag1', 'tag2']).catch(() => {
      expect(store.dispatch.mock.calls.length).toBe(1);
      expect(store.dispatch).toHaveBeenCalledWith(
        changeProgramTags(1, ['tag1', 'tag2'], {
          headers: {
            Authorization: `JWT ${cookiesValues.auth_jwt}`,
          },
        }),
      );
      done();
    });
  });
});
