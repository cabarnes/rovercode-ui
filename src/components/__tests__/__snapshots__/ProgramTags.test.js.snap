// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ProgramTags component renders on the page with no errors 1`] = `
<withCookies(injectIntl(Connect(ProgramTags)))
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "payload": Promise {},
              "type": "FETCH_TAGS",
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <withCookies(injectIntl(Connect(ProgramTags)))
    cookies={
      Cookies {
        "HAS_DOCUMENT_COOKIE": true,
        "changeListeners": Array [
          [Function],
        ],
        "cookies": Object {},
      }
    }
    forwardedRef={null}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [MockFunction] {
          "calls": Array [
            Array [
              Object {
                "payload": Promise {},
                "type": "FETCH_TAGS",
              },
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": Promise {},
            },
          ],
        },
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <injectIntl(Connect(ProgramTags))
      allCookies={Object {}}
      cookies={
        Cookies {
          "HAS_DOCUMENT_COOKIE": true,
          "changeListeners": Array [
            [Function],
          ],
          "cookies": Object {},
        }
      }
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [MockFunction] {
            "calls": Array [
              Array [
                Object {
                  "payload": Promise {},
                  "type": "FETCH_TAGS",
                },
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": Promise {},
              },
            ],
          },
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <Connect(ProgramTags)
        allCookies={Object {}}
        cookies={
          Cookies {
            "HAS_DOCUMENT_COOKIE": true,
            "changeListeners": Array [
              [Function],
            ],
            "cookies": Object {},
          }
        }
        intl={
          Object {
            "defaultFormats": Object {},
            "defaultLocale": "en",
            "formatDate": [Function],
            "formatDateToParts": [Function],
            "formatDisplayName": [Function],
            "formatHTMLMessage": [Function],
            "formatList": [Function],
            "formatMessage": [Function],
            "formatNumber": [Function],
            "formatNumberToParts": [Function],
            "formatPlural": [Function],
            "formatRelativeTime": [Function],
            "formatTime": [Function],
            "formatTimeToParts": [Function],
            "formats": Object {},
            "formatters": Object {
              "getDateTimeFormat": [Function],
              "getDisplayNames": [Function],
              "getListFormat": [Function],
              "getMessageFormat": [Function],
              "getNumberFormat": [Function],
              "getPluralRules": [Function],
              "getRelativeTimeFormat": [Function],
            },
            "locale": "en",
            "messages": Object {},
            "onError": [Function],
            "textComponent": Symbol(react.fragment),
            "timeZone": undefined,
          }
        }
        store={
          Object {
            "clearActions": [Function],
            "dispatch": [MockFunction] {
              "calls": Array [
                Array [
                  Object {
                    "payload": Promise {},
                    "type": "FETCH_TAGS",
                  },
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": Promise {},
                },
              ],
            },
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
        }
      >
        <ProgramTags
          allCookies={Object {}}
          changeProgramTags={[Function]}
          code={
            Object {
              "isReadOnly": false,
              "tags": Array [
                "tag1",
                "tag2",
              ],
            }
          }
          cookies={
            Cookies {
              "HAS_DOCUMENT_COOKIE": true,
              "changeListeners": Array [
                [Function],
              ],
              "cookies": Object {},
            }
          }
          fetchTags={[Function]}
          intl={
            Object {
              "defaultFormats": Object {},
              "defaultLocale": "en",
              "formatDate": [Function],
              "formatDateToParts": [Function],
              "formatDisplayName": [Function],
              "formatHTMLMessage": [Function],
              "formatList": [Function],
              "formatMessage": [Function],
              "formatNumber": [Function],
              "formatNumberToParts": [Function],
              "formatPlural": [Function],
              "formatRelativeTime": [Function],
              "formatTime": [Function],
              "formatTimeToParts": [Function],
              "formats": Object {},
              "formatters": Object {
                "getDateTimeFormat": [Function],
                "getDisplayNames": [Function],
                "getListFormat": [Function],
                "getMessageFormat": [Function],
                "getNumberFormat": [Function],
                "getPluralRules": [Function],
                "getRelativeTimeFormat": [Function],
              },
              "locale": "en",
              "messages": Object {},
              "onError": [Function],
              "textComponent": Symbol(react.fragment),
              "timeZone": undefined,
            }
          }
          store={
            Object {
              "clearActions": [Function],
              "dispatch": [MockFunction] {
                "calls": Array [
                  Array [
                    Object {
                      "payload": Promise {},
                      "type": "FETCH_TAGS",
                    },
                  ],
                ],
                "results": Array [
                  Object {
                    "type": "return",
                    "value": Promise {},
                  },
                ],
              },
              "getActions": [Function],
              "getState": [Function],
              "replaceReducer": [Function],
              "subscribe": [Function],
            }
          }
          tag={
            Object {
              "tags": Array [
                Object {
                  "name": "tag1",
                },
                Object {
                  "name": "tag2",
                },
                Object {
                  "name": "tag3",
                },
              ],
            }
          }
        >
          <Grid
            centered={true}
            container={true}
          >
            <div
              className="ui centered container grid"
            >
              <GridRow>
                <div
                  className="row"
                >
                  <Dropdown
                    additionLabel="Add "
                    additionPosition="top"
                    allowAdditions={true}
                    closeOnBlur={true}
                    closeOnChange={true}
                    closeOnEscape={true}
                    deburr={false}
                    disabled={false}
                    fluid={true}
                    icon="dropdown"
                    minCharacters={1}
                    multiple={true}
                    noResultsMessage="No results found."
                    onAddItem={[Function]}
                    onChange={[Function]}
                    openOnFocus={true}
                    options={
                      Array [
                        Object {
                          "key": "tag1",
                          "text": "tag1",
                          "value": "tag1",
                        },
                        Object {
                          "key": "tag2",
                          "text": "tag2",
                          "value": "tag2",
                        },
                        Object {
                          "key": "tag3",
                          "text": "tag3",
                          "value": "tag3",
                        },
                      ]
                    }
                    placeholder="Add tags..."
                    renderLabel={[Function]}
                    search={true}
                    searchInput="text"
                    selectOnBlur={true}
                    selectOnNavigation={true}
                    selection={true}
                    value={
                      Array [
                        "tag1",
                        "tag2",
                      ]
                    }
                    wrapSelection={true}
                  >
                    <Ref
                      innerRef={
                        Object {
                          "current": <div
                            aria-disabled="false"
                            aria-expanded="false"
                            class="ui fluid multiple search selection dropdown"
                            role="combobox"
                          >
                            <a
                              class="ui label"
                              value="tag1"
                            >
                              tag1
                              <i
                                aria-hidden="true"
                                class="delete icon"
                              />
                            </a>
                            <a
                              class="ui label"
                              value="tag2"
                            >
                              tag2
                              <i
                                aria-hidden="true"
                                class="delete icon"
                              />
                            </a>
                            <input
                              aria-autocomplete="list"
                              autocomplete="off"
                              class="search"
                              tabindex="0"
                              type="text"
                              value=""
                            />
                            <span
                              class="sizer"
                            />
                            <div
                              aria-atomic="true"
                              aria-live="polite"
                              class="text"
                              role="alert"
                            />
                            <i
                              aria-hidden="true"
                              class="dropdown icon"
                            />
                            <div
                              aria-multiselectable="true"
                              class="menu transition"
                              role="listbox"
                            >
                              <div
                                aria-checked="false"
                                aria-selected="true"
                                class="selected item"
                                role="option"
                                style="pointer-events: all;"
                              >
                                <span
                                  class="text"
                                >
                                  tag3
                                </span>
                              </div>
                            </div>
                          </div>,
                        }
                      }
                    >
                      <RefFindNode
                        innerRef={
                          Object {
                            "current": <div
                              aria-disabled="false"
                              aria-expanded="false"
                              class="ui fluid multiple search selection dropdown"
                              role="combobox"
                            >
                              <a
                                class="ui label"
                                value="tag1"
                              >
                                tag1
                                <i
                                  aria-hidden="true"
                                  class="delete icon"
                                />
                              </a>
                              <a
                                class="ui label"
                                value="tag2"
                              >
                                tag2
                                <i
                                  aria-hidden="true"
                                  class="delete icon"
                                />
                              </a>
                              <input
                                aria-autocomplete="list"
                                autocomplete="off"
                                class="search"
                                tabindex="0"
                                type="text"
                                value=""
                              />
                              <span
                                class="sizer"
                              />
                              <div
                                aria-atomic="true"
                                aria-live="polite"
                                class="text"
                                role="alert"
                              />
                              <i
                                aria-hidden="true"
                                class="dropdown icon"
                              />
                              <div
                                aria-multiselectable="true"
                                class="menu transition"
                                role="listbox"
                              >
                                <div
                                  aria-checked="false"
                                  aria-selected="true"
                                  class="selected item"
                                  role="option"
                                  style="pointer-events: all;"
                                >
                                  <span
                                    class="text"
                                  >
                                    tag3
                                  </span>
                                </div>
                              </div>
                            </div>,
                          }
                        }
                      >
                        <div
                          aria-disabled={false}
                          aria-expanded={false}
                          className="ui fluid multiple search selection dropdown"
                          onBlur={[Function]}
                          onChange={[Function]}
                          onClick={[Function]}
                          onFocus={[Function]}
                          onMouseDown={[Function]}
                          role="combobox"
                        >
                          <Label
                            active={false}
                            as="a"
                            content="tag1"
                            key="tag1"
                            onClick={[Function]}
                            onRemove={[Function]}
                            value="tag1"
                          >
                            <a
                              className="ui label"
                              onClick={[Function]}
                              value="tag1"
                            >
                              tag1
                              <Icon
                                as="i"
                                name="delete"
                                onClick={[Function]}
                              >
                                <i
                                  aria-hidden="true"
                                  className="delete icon"
                                  onClick={[Function]}
                                />
                              </Icon>
                            </a>
                          </Label>
                          <Label
                            active={false}
                            as="a"
                            content="tag2"
                            key="tag2"
                            onClick={[Function]}
                            onRemove={[Function]}
                            value="tag2"
                          >
                            <a
                              className="ui label"
                              onClick={[Function]}
                              value="tag2"
                            >
                              tag2
                              <Icon
                                as="i"
                                name="delete"
                                onClick={[Function]}
                              >
                                <i
                                  aria-hidden="true"
                                  className="delete icon"
                                  onClick={[Function]}
                                />
                              </Icon>
                            </a>
                          </Label>
                          <Ref
                            innerRef={
                              Object {
                                "current": <input
                                  aria-autocomplete="list"
                                  autocomplete="off"
                                  class="search"
                                  tabindex="0"
                                  type="text"
                                  value=""
                                />,
                              }
                            }
                          >
                            <RefFindNode
                              innerRef={
                                Object {
                                  "current": <input
                                    aria-autocomplete="list"
                                    autocomplete="off"
                                    class="search"
                                    tabindex="0"
                                    type="text"
                                    value=""
                                  />,
                                }
                              }
                            >
                              <DropdownSearchInput
                                autoComplete="off"
                                key="text"
                                onChange={[Function]}
                                style={
                                  Object {
                                    "width": undefined,
                                  }
                                }
                                tabIndex={0}
                                type="text"
                                value=""
                              >
                                <input
                                  aria-autocomplete="list"
                                  autoComplete="off"
                                  className="search"
                                  onChange={[Function]}
                                  style={
                                    Object {
                                      "width": undefined,
                                    }
                                  }
                                  tabIndex={0}
                                  type="text"
                                  value=""
                                />
                              </DropdownSearchInput>
                            </RefFindNode>
                          </Ref>
                          <span
                            className="sizer"
                          />
                          <div
                            aria-atomic={true}
                            aria-live="polite"
                            className="text"
                            role="alert"
                          />
                          <Icon
                            as="i"
                            className=""
                            name="dropdown"
                            onClick={[Function]}
                          >
                            <i
                              aria-hidden="true"
                              className="dropdown icon"
                              onClick={[Function]}
                            />
                          </Icon>
                          <DropdownMenu
                            aria-multiselectable={true}
                            role="listbox"
                          >
                            <div
                              aria-multiselectable={true}
                              className="menu transition"
                              role="listbox"
                            >
                              <DropdownItem
                                active={false}
                                key="tag3"
                                onClick={[Function]}
                                selected={true}
                                style={
                                  Object {
                                    "pointerEvents": "all",
                                  }
                                }
                                text="tag3"
                                value="tag3"
                              >
                                <div
                                  aria-checked={false}
                                  aria-selected={true}
                                  className="selected item"
                                  onClick={[Function]}
                                  role="option"
                                  style={
                                    Object {
                                      "pointerEvents": "all",
                                    }
                                  }
                                >
                                  <span
                                    className="text"
                                  >
                                    tag3
                                  </span>
                                </div>
                              </DropdownItem>
                            </div>
                          </DropdownMenu>
                        </div>
                      </RefFindNode>
                    </Ref>
                  </Dropdown>
                </div>
              </GridRow>
              <GridRow>
                <div
                  className="row"
                >
                  <Button
                    as="button"
                    disabled={true}
                    onClick={[Function]}
                    primary={true}
                  >
                    <Ref
                      innerRef={
                        Object {
                          "current": <button
                            class="ui primary disabled button"
                            disabled=""
                            tabindex="-1"
                          >
                            Save
                          </button>,
                        }
                      }
                    >
                      <RefFindNode
                        innerRef={
                          Object {
                            "current": <button
                              class="ui primary disabled button"
                              disabled=""
                              tabindex="-1"
                            >
                              Save
                            </button>,
                          }
                        }
                      >
                        <button
                          className="ui primary disabled button"
                          disabled={true}
                          onClick={[Function]}
                          tabIndex={-1}
                        >
                          <FormattedMessage
                            defaultMessage="Save"
                            id="app.program_tags.save"
                            values={Object {}}
                          >
                            Save
                          </FormattedMessage>
                        </button>
                      </RefFindNode>
                    </Ref>
                  </Button>
                </div>
              </GridRow>
            </div>
          </Grid>
        </ProgramTags>
      </Connect(ProgramTags)>
    </injectIntl(Connect(ProgramTags))>
  </withCookies(injectIntl(Connect(ProgramTags)))>
</withCookies(injectIntl(Connect(ProgramTags)))>
`;
